### PR TITLE
Add vertical alignment to blocks using flex control

### LINKED ISSUE #38498: Group Block [Row]: Support both vertical and horizontal alignment
## What problem does this address?

Right now using the row block the only vertical alignment possible is centered. Layouts like the one from the screenshot require extra CSS.

Screenshot 2022-02-03 at 17 28 08

## What is your proposed solution?

Since we are using flexbox, we could use vertical alignments like the column and the media+text blocks do:

Screenshot 2022-02-03 at 17 57 34

I've had a look and both the columns block and the media+text use `align-self` instead of `align-items`, I wonder what would be the best case for this.

I'd love your thoughts on this @youknowriad

### PR DESCRIPTION
**Pull request by JustinyAhin**
Fixes https://github.com/WordPress/gutenberg/issues/38498

## What?
This PR adds a new block control for vertical alignment to the blocks that use flex control.

## Why?
See https://github.com/WordPress/gutenberg/issues/38498

## How?
This adds a new `verticalAlignment` (https://github.com/WordPress/gutenberg/blob/190a207971655295ba8efffd9805cac1768fae84/packages/block-editor/src/components/block-vertical-alignment-control/index.js#L17) to the group block, and the according `align-items` styles. 

## Testing Instructions
1. In a new post, add a new Row block
2. Add some content to the block
3. Set some padding to the block and with they increase the block height (by for instance adding a `height` property from the browser inspector)
4. Add some background color to the block to make the changes visible
5. Change the "Vertical alignment" values in the block control toolbar

### CODE REVIEW

File: packages/block-library/src/group/edit.js [hunk 1]
```diff
@@ -68,6 +90,12 @@ function GroupEdit( { attributes, setAttributes, clientId } ) {
 
 	return (
 		<>
+			<BlockControls>
+				<BlockVerticalAlignmentControl
```
Comment (youknowriad) [inline packages/block-library/src/group/edit.js]: Is there any reason for this to be specific to the group block? Feels like something generic that can be used by any "flex" layout. (So we can move its implementation details both edit/save to `flex.js` file).

**Review by youknowriad — COMMENTED** [review]
Thanks for working on this feature

**Comment by youknowriad** [comment]
Thanks for the update, this looking better code wise, let's try to get some reviews on the design/behavior aspect.

**Comment by MaggieCabrera** [comment]
Doesn't seem to be working for me on a row block using the theme Remote (https://wordpress.org/themes/remote/)

Screen Capture on 2022-04-05 at 11-40-05

**Comment by JustinyAhin** [comment]
> Thanks for the update, this looking better code wise, let's try to get some reviews on the design/behavior aspect.

@youknowriad there are still some adjustments to do in the `save()` function, I believe.

**Comment by jasmussen** [comment]
Nice, thanks for the PR, took a quick spin playing around with the vertical toolbar alignment controls:
row vertical alignment

High level, this will be a very useful feature to have. It also feels worthwhile to have a button present in the block toolbar as it's valuable.

As shown in the GIF, I couldn't get the alignment to actually work — I tested a paragraph and an imagea, and the paragraph kept vertically centered. Did I do anything wrong?

Some thoughts:

* Without a way to set a min-height on the Stack variation, does this control make sense there?
* Once we _do_ get min-height controls for Stack and Row blocks (perhaps even group, see also #28845) — `space-between` would be a nice option to have even for both these variations.
* Should we default to top vertical alignment?

Finally, the Layout panel likely showing horizontal alignment but not vertical stands out as a little uncanny. It would be nice to find a design that could accommodate both of these. This is an area where Figma might be on to something, as it unifies both horizontal and vertical alignments into a single unified panel:
Screenshot 2022-04-05 at 11 44 26
Screenshot 2022-04-05 at 11 44 31

This is not unlike what we already have for the Cover block:

Screenshot 2022-04-05 at 11 45 23

In other words, there's an opportunity there to simplify. The larger question is whether that needs to block this PR. Probably not?

**Comment by JustinyAhin** [comment]
@jasmussen @MaggieCabrera with https://github.com/WordPress/gutenberg/pull/40013/commits/d6c92538a03dce8b4cd4d752382b593af24fc504 you should be able to make it works.

File: packages/block-editor/src/layouts/flex.js [hunk 2]
```diff
@@ -113,9 +119,11 @@ export default {
 		const flexWrap = flexWrapOptions.includes( layout.flexWrap )
 			? layout.flexWrap
 			: 'wrap';
+		const verticalAlignment =
+			verticalAlignmentMap[ layout.verticalAlignment ];
```
Comment (ntsekouras) [inline packages/block-editor/src/layouts/flex.js]: You need to add a fallback to previous `center` value.

Comment (ntsekouras) [reply packages/block-editor/src/layouts/flex.js]: Also the respective default for vertical orientation was the browser default so there was no visual issue, but in this case the `center` is not the default and we need to make the button active probably.

File: packages/block-editor/src/layouts/flex.js [hunk 3]
```diff
@@ -36,6 +40,12 @@ const alignItemsMap = {
 	center: 'center',
 };
 
+const verticalAlignmentMap = {
+	top: 'flex-start',
+	middle: 'center',
```
Comment (aristath) [inline packages/block-editor/src/layouts/flex.js]: shouldn't this be like this?
```suggestion
	center: 'center',
```

File: packages/block-editor/src/layouts/flex.js (hunk 2 above)
Comment (JustinyAhin) [reply packages/block-editor/src/layouts/flex.js]: > Also the respective default for vertical orientation was the browser default so there was no visual issue, but in this 
case the center is not the default and we need to make the button active probably.

I didn't really understand this part, could you please explain me in detail?

File: packages/block-editor/src/layouts/flex.js [hunk 4]
```diff
@@ -140,6 +158,62 @@ export default {
 	},
 };
 
+function FlexLayoutVerticalAlignmentControl( {
+	layout,
+	onChange,
+	isToolbar = false,
+} ) {
+	const { verticalAlignment = 'top' } = layout;
```
Comment (ntsekouras) [inline packages/block-editor/src/layouts/flex.js]:
```suggestion
	const { verticalAlignment = verticalAlignmentMap.center } = layout;

```

File: packages/block-editor/src/layouts/flex.js (hunk 2 above)
Comment (ntsekouras) [reply packages/block-editor/src/layouts/flex.js]: It seems we just need [this change](https://github.com/WordPress/gutenberg/pull/40013/files#r842765973). Disregard my mention to browser defaults, because code might have changed here 😄 .

The point was that the controls should match the default/fallback value you added.

**Review by aristath — APPROVED** [review]
Great work @JustinyAhin, thank you for working on this one! ❤️ 🎉 

The code looks good and makes sense.
Tested it and works as advertised. 👍

File: packages/block-editor/src/layouts/flex.js [hunk 5]
```diff
@@ -140,6 +158,63 @@ export default {
...
+		},
+		{
+			value: 'center',
+			label: __( 'Align items center' ),
+		},
+		{
+			value: 'flex-end',
+			label: __( 'Align items bottom' ),
+		},
+	];
```
Comment (mikachan) [inline packages/block-editor/src/layouts/flex.js:200]: I think these options need to be replicated in [lib/block-supports/layout.php](https://github.com/WordPress/gutenberg/blob/trunk/lib/block-supports/layout.php#L80), similar to how the `$justify_content_options` array works. For example, by adding a vertical alignment options array:

```
$vertical_alignment_options = array(
	'top'    => 'flex-start',
	'center' => 'center',
	'bottom' => 'flex-end'
);
```

And then using these options, with `$layout['verticalAlignment']`, around [line 111](https://github.com/WordPress/gutenberg/blob/trunk/lib/block-supports/layout.php#L111), inside the `$layout_orientation` check:

```
if ( ! empty( $layout['verticalAlignment'] ) && array_key_exists( $layout['verticalAlignment'], $vertical_alignment_options ) ) {
	$style .= "align-items: {$vertical_alignment_options[ $layout['verticalAlignment'] ]};";
} else {
	$style .= 'align-items: center;';
}
```

**Review by mikachan — COMMENTED** [review]
Thanks for working on this! It's working well for me in the editor, however, on the front end I don't seem to see the vertical alignment being updated:

Editor:
image

Front end:
image

I've left another comment with a possible solution.

File: packages/block-editor/src/layouts/flex.js (hunk 5 above)
Comment (JustinyAhin) [reply packages/block-editor/src/layouts/flex.js:200]: Thanks @mikachan.

**Review by mikachan — APPROVED** [review]
Those changes look good to me, this is working great on the front end now as well 🎉

**Comment by carolinan** [comment]
What is the expected result for the navigation block which uses the flex control?

I would like to be able to align the text items with the top of the logo, for example.
Here is a navigation block with mixed items, (placed in the post editor without containers):

https://user-images.githubusercontent.com/7422055/162396948-9551db45-e9c6-4a21-a668-bf01ae7a4ada.mp4
(Nothing happens when you use the vertical alignment control on the navigation block)

**Comment by carolinan** [comment]
Buttons work well:

https://user-images.githubusercontent.com/7422055/162398248-1e95c97b-d6cf-4def-bfee-f8235f6ed40d.mp4
3 buttons placed in the post editor: The middle button is larger, and the two smaller buttons align to the top, middle and bottom of the larger button correctly.

**Comment by carolinan** [comment]
Query pagination, comment query pagination -Works in the editor and front.
Columns, column -I do not see any conflicts with the existing settings.
Social links: The icons are all the same size, so the vertical alignment does nothing.

**Comment by aristath** [comment]
I added a condition that will only render the control when using the horizontal alignment, and implemented a new `allowVerticalAlignment` flag so we can disable the feature for blocks that don't work with this.
As soon as tests pass I think it's safe to merge this one, and in a subsequent PR disable verticalAlignment in the navigation and social blocks @carolinan 👍
